distributed/worker_state_machine.py

```diff
@@ -1808,19 +1815,33 @@ def _put_key_in_memory(
         """
         if ts.key in self.data:
             ts.state = "memory"
-            return {}
+            return {}, []
 
         recommendations: Recs = {}
+        instructions: Instructions = []
+
         if ts.key in self.actors:
             self.actors[ts.key] = value
         else:
             start = time()
             self.data[ts.key] = value
             stop = time()
-            if stop - start > 0.020:
+            if stop - start > 0.005:
                 ts.startstops.append(
                     {"action": "disk-write", "start": start, "stop": stop}
                 )
+                instructions.append(
+                    DigestMetric(
```

Comment:
Why does this need to be an instruction? I get the purity of the state machine not having side effects, but functionally, does it make any difference whether we emit an instruction vs just do the digest update when it should happen?

When we move to typical use of the Prometheus client https://github.com/dask/distributed/issues/7364#issuecomment-1352411643, I expect we'd just call `disk_write_target_duration.observe(stop - start)` here. The instruction seems like a bit of indirection.

Reply:
Because it's a clean way to call `Server.digest_metric` from the state machine. The alternatives were (a) have a reverse weakref to the Worker or (b) have a Worker global variable. Both of which taint the state machine and introduce special code paths that are not executed when it's run in isolation in many unit tests.

Reply:
Makes sense, thanks for explaining. I forgot the class hierarchy of the state machine worked that way.

So it seems like after https://github.com/dask/distributed/issues/7364, we would be able to update the metrics directly in the state machine without an instruction, but for now we need that to bridge the StateMachine<->Server gap.